system/lib/pthread/library_pthread.c

```diff
@@ -644,7 +505,18 @@ void emscripten_main_thread_process_queued_calls() {
   if (!_emscripten_allow_main_runtime_queued_calls)
     return;
 
+  // Recursion guard to avoid infinite recursion when we arrive here from the
+  // pthread_lock calls inside `emscripten_proxy_execute_queue`. This isn't
+  // caught by the queue's own recursion guard because the lock has to be
```

Comment:
> "This isn't caught by the queue's own recursion guard because the lock has to be
 acquired before that recursion guard can be checked"

This seems a little worrying, can we so something special for this internal lock to avoid the recursion.

Protecting against it here seems like treating the symptom rather than the cause..  for example what happens if a user calls `emscripten_current_thread_process_queued_calls` directly on the main thread instead of going via this wrapper?

Reply:
I moved the recursion guard into `emscripten_proxy_execute_queue`, so at least it's more robust now and avoids the problem you identified.